MoreLinq/DeconstructibleEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+
+    /// <summary>
+    /// Represents a source whose elements can be sequentially deconstructed
+    /// into separate variables.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements.</typeparam>
+
+    public partial struct DeconstructibleEnumerable<T> : IEnumerable<T>
+    {
+        readonly IEnumerable<T> _source;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DeconstructibleEnumerable{T}"/>
+        /// with a sequence.
+        /// </summary>
+        /// <param name="source">
+        /// The source sequence on which to enable deconstruction.</param>
+
+        public DeconstructibleEnumerable(IEnumerable<T> source) =>
```

Comment:
Is it necessary to make this public? It could be internal, and then we expose less.

Reply:
No and that's how I had it but I often feel that the one shouldn't lie about using a `struct`. Do we gain anything by blocking the use of this constructor?

Reply:
Mostly freedom. I've usually thought it is better to offer the minimum visibility that is required. But I'm not sure it matters much in this case, as I don't think the constructor will be modified much.

Reply:
> I've usually thought it is better to offer the minimum visibility that is required. 

I'm in that camp too, but technically speaking, this type stands on its own feet and does exactly what it advertises. I'm just playing the devil's advocate here. Will give it some more thought before merging the PR provided it's not a basis for your disapproval of the design.

Reply:
This is not a basis for disapproval, but I wondered if there was any need to make it public

Reply:
…because I considered it [mostly harmless](https://en.wikipedia.org/wiki/Mostly_Harmless). 😄